miasm2/core/asmblock.py

```diff
@@ -359,132 +323,209 @@ class AsmSymbolPool(object):
+
+        self._name_to_loc_key = {}
+        self._offset_to_loc_key = {}
+
+        self._loc_key_num = 0
 
-    def add_label(self, name, offset=None):
+    def loc_key_to_offset(self, loc_key):
         """
-        Create and add a label to the symbol_pool
-        @name: label's name
-        @offset: (optional) label's offset
+        Return offset of @loc_key, None else.
+        @loc_key: LocKey instance
         """
-        label = AsmLabel(name, offset)
+        return self._loc_key_to_offset.get(loc_key, None)
+
+    def loc_key_to_name(self, loc_key):
+        """
+        Return name of @loc_key.
+        @loc_key: LocKey instance
+        """
+        return self._loc_key_to_name.get(loc_key)
+
+    def add_location(self, name, offset=None):
+        """
+        Create and add a location to the symbol_pool
+        @name: loc_key's name (never empty)
+        @offset: (optional) loc_key's offset
+        """
+
+        if is_int(name):
+            assert offset is None or offset == name
+            offset = name
+            name = "loc_%.16X" % (int(name) & 0xFFFFFFFFFFFFFFFF)
+        if offset is not None:
+            offset = int(offset)
+
+        assert name
+        assert not is_int(name)
 
         # Test for collisions
-        if (label.offset in self._offset2label and
-                label != self._offset2label[label.offset]):
-            raise ValueError('symbol %s has same offset as %s' %
-                             (label, self._offset2label[label.offset]))
-        if (label.name in self._name2label and
-                label != self._name2label[label.name]):
-            raise ValueError('symbol %s has same name as %s' %
-                             (label, self._name2label[label.name]))
-
-        self._labels.add(label)
-        if label.offset is not None:
-            self._offset2label[label.offset] = label
-        if label.name != "":
-            self._name2label[label.name] = label
-        return label
-
-    def remove_label(self, label):
+        known_loc_key = self.getby_name(name)
+        if known_loc_key is not None:
+            known_offset = self.loc_key_to_offset(known_loc_key)
+            if known_offset != offset:
+                raise ValueError(
+                    'symbol %s with different offset %s %s' % (
+                        name, offset, known_offset
+                    )
+                )
+            return known_loc_key
+
+        if offset is not None:
+            known_loc_key = self.getby_offset(offset)
+            if known_loc_key:
+                known_name = self.loc_key_to_name(known_loc_key)
+                if known_name != name:
+                    raise ValueError(
+                        'offset %s with different names %s %s' % (
+                            offset, name, known_name
+                        )
+                    )
+                return known_loc_key
+
+        loc_key = LocKey(self._loc_key_num)
```

Comment:
This may be a good thing that `LocKey` generation is done only in one place: `gen_loc_key`, and then this last method call from there

Reply:
not a bug